session/portforward.go

```diff
@@ -0,0 +1,276 @@
+					Properties: map[string]interface{}{
+						wspf.PropertyPortForwardConnectionID: f.ConnectionID,
+					},
+				},
+			}
+			if err := f.ResponseWriter.WriteProtoMsg(m); err != nil {
+				log.Errorf("portForwardHandler: webSock.WriteMessage(%+v)", err)
+			}
+			// close the forwarder
+			f.Close()
+		case data := <-dataChan:
+			log.Debugf("port-forward[%s/%s] read %d bytes", f.SessionID, f.ConnectionID, len(data))
+
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ws.ProtoTypePortForward,
+					MsgType:   wspf.MessageTypePortForward,
+					SessionID: f.SessionID,
+					Properties: map[string]interface{}{
+						wspf.PropertyPortForwardConnectionID: f.ConnectionID,
+					},
+				},
+				Body: data,
+			}
+			if err := f.ResponseWriter.WriteProtoMsg(m); err != nil {
+				log.Errorf("portForwardHandler: webSock.WriteMessage(%+v)", err)
+			}
+		case <-f.ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *MenderPortForwarder) Write(body []byte) error {
+	log.Debugf("port-forward[%s/%s] write %d bytes", f.SessionID, f.ConnectionID, len(body))
+	_, err := f.conn.Write(body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func PortForward() Constructor {
+	f := HandlerFunc(portForwardHandler)
+	return func() SessionHandler { return f }
+}
+
+func portForwardHandler(msg *ws.ProtoMsg, w ResponseWriter) {
+	var err error
+	switch msg.Header.MsgType {
+	case wspf.MessageTypePortForwardNew:
+		err = portForwardHandlerNew(msg, w)
+	case wspf.MessageTypePortForwardStop:
+		err = portForwardHandlerStop(msg, w)
+	case wspf.MessageTypePortForward:
+		err = portForwardHandlerForward(msg, w)
+	default:
+		err = errPortForwardUnkonwnMessageType
+	}
+	if err != nil {
+		log.Errorf("portForwardHandler(%+v)", err)
+
+		response := &ws.ProtoMsg{
+			Header: ws.ProtoHdr{
+				Proto:     ws.ProtoTypeControl,
+				MsgType:   ws.MessageTypeError,
+				SessionID: msg.Header.SessionID,
+			},
+			Body: []byte(err.Error()),
+		}
+		if err := w.WriteProtoMsg(response); err != nil {
+			log.Errorf("portForwardHandler: webSock.WriteMessage(%+v)", err)
+		}
+	}
+}
+
+func getConnectionIDFromMessage(message *ws.ProtoMsg) string {
+	connectionID, _ := message.Header.Properties[wspf.PropertyPortForwardConnectionID].(string)
+	return connectionID
+}
+
+func portForwardHandlerNew(message *ws.ProtoMsg, w ResponseWriter) error {
+	connectionID := getConnectionIDFromMessage(message)
+	protocol, _ := message.Header.Properties[wspf.PropertyPortForwardProtocol].(string)
+	host, _ := message.Header.Properties[wspf.PropertyPortForwardRemoteHost].(string)
+	portNumber, _ := message.Header.Properties[wspf.PropertyPortForwardRemotePort].(uint64)
+
+	if connectionID == "" || portNumber == 0 || protocol == "" {
+		return errPortForwardInvalidMessage
+	}
+
+	portForwarder := &MenderPortForwarder{
+		ConnectionID:   connectionID,
+		SessionID:      message.Header.SessionID,
+		ResponseWriter: w,
+	}
+
+	key := message.Header.SessionID + "/" + connectionID
```

Comment:
Not necessary to use the SessionID as it will be the same for one instance of this handler.